test/Microsoft.AspNetCore.Authentication.Test/Cookies/CookieMiddlewareTests.cs

```diff
@@ -713,6 +715,155 @@ private Task SignOutAsWrong(HttpContext context)
+            var message = new HttpRequestMessage(new HttpMethod(httpMethod), "http://example.com/requires-authentication");
+            var transaction2 = await SendAsync(server, message, transaction1.CookieNameValue);
+            Assert.Equal("Alice", FindClaimValue(transaction2, ClaimTypes.Name));
+        }
+
+        // These http methods are allow-listed
+        [Theory]
+        [InlineData("GET")]
+        [InlineData("HEaD")]
+        [InlineData("TRace")]
+        [InlineData("options")]
+        public async Task Cookie_Authenticated_WithCSRF(string httpMethod)
+        {
+            var builder = new WebHostBuilder()
+                .ConfigureServices(services =>
+                {
+                    services.AddAntiforgery(options => options.FormFieldName = "CSRF_REQUEST_TOKEN");
+                    services.AddAuthentication();
+                })
+                .Configure(app =>
+                {
+                    app.UseCookieAuthentication();
+
+                    app.Run(async context =>
+                    {
+                        if (context.Request.Path == new PathString("/login"))
+                        {
+                            await SignInAsAlice(context);
+
+                            var antiforgery = context.RequestServices.GetRequiredService<IAntiforgery>();
+
+                            var tokens = antiforgery.GetAndStoreTokens(context);
+                            context.Response.Headers["CSRF"] = tokens.RequestToken;
+                        }
+                        else if (context.Request.Path == new PathString("/requires-authentication"))
+                        {
+                            var authorizationContext = new AuthenticateContext(CookieAuthenticationDefaults.AuthenticationScheme);
+                            await context.Authentication.AuthenticateAsync(authorizationContext);
+                            Describe(context.Response, authorizationContext);
+                        }
+                    });
+                });
+
+            var server = new TestServer(builder);
+
+            var transaction1 = await SendAsync(server, "http://example.com/login");
+
+            var message = new HttpRequestMessage(HttpMethod.Post, "http://example.com/requires-authentication");
+            message.Content = new FormUrlEncodedContent(new Dictionary<string, string>()
+            {
+                { "CSRF_REQUEST_TOKEN", transaction1.Response.Headers.GetValues("CSRF").First() },
+            });
+
+            var transaction2 = await SendAsync(server, message, transaction1.CookieNameValue);
+            Assert.Equal("Alice", FindClaimValue(transaction2, ClaimTypes.Name));
+        }
+
+        [Theory]
+        [InlineData("PUT")]
+        [InlineData("PoST")]
+        [InlineData("delete")]
+        [InlineData("DEBUG")]
+        [InlineData("bleh")]
+        public async Task Cookie_TreatedAsAnonymous_WithoutCSRF(string httpMethod)
```

Comment:
Just a reminder for myself that I'm missing a test for:
- Authenticated + POST + has CSRF
- CSRF protection suppressed by options